src/mrack/providers/provider.py

```diff
@@ -30,16 +37,116 @@ def name(self):
 
     async def validate_hosts(self, hosts):
         """Validate that host requirements are well specified."""
-        return
+        raise NotImplementedError()
 
     async def can_provision(self, hosts):
         """Check that provider has enough resources to provison hosts."""
         raise NotImplementedError()
 
-    async def provision_hosts(self, hosts):
-        """Provision hosts and wait for it to finish."""
+    async def create_server(self, req):
+        """Request and create resource on selected provider."""
+        raise NotImplementedError()
+
+    async def wait_till_provisioned(self, resource):
+        """Wait till resource is provisioned."""
+        raise NotImplementedError()
+
+    async def provision_hosts(self, hosts, logger=None):
+        """Provision hosts based on list of host requirements.
+
+        Main provider method for provisioning.
+
+        First it validates that host requirements are valid and that
+        provider has enough resources(quota).
+
+        Then issues provisioning and waits for it succeed. Raises exception if any of
+        the servers was not successfully provisioned. If that happens it issues deletion
+        of all already provisioned resources.
+
+        Return list of information about provisioned servers.
+        """
+        if not logger:
+            logger = logging.getLogger(self.name)
```

Comment:
I guess you want to pass the `logger` instance to have a 'proper' prefix, right? I _think_ this goes against some of the assumptions the logging makes.